File: Java/146 LRU Cache.java
class LRUCache {
     int capacity;
     Map<Integer, Node> mapToPrev;
     Node dummy;
     Node tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.mapToPrev = new HashMap<Integer, Node>();
        this.dummy = new Node(0,0);
        this.tail = dummy;
    }
    
    public int get(int key) {
        if(!mapToPrev.containsKey(key))
            return -1;
        kick(key);
        return tail.val;
    }
    
    public void put(int key, int value) {
        if(mapToPrev.containsKey(key)){
            kick(key);
            tail.val = value;
        }else{
            Node newNode = new Node(key, value);
            pushBack(newNode);
            if(mapToPrev.size() > capacity){
                popFront();
            }
        }
    }
    
    private void pushBack(Node node){
        mapToPrev.put(node.key, tail);
        tail.next = node;
        tail = node;
    }
    
	//kick the key node to the tail
    private void kick(int key){
        Node prevNode = mapToPrev.get(key);
        Node currNode = prevNode.next;
        if(currNode == tail){
            return;
        }
        prevNode.next = currNode.next;
        mapToPrev.put(prevNode.next.key,prevNode);
        currNode.next = null;
        pushBack(currNode);
    }
    
    private void popFront(){
        Node head = dummy.next;
        mapToPrev.remove(head.key);
        mapToPrev.put(head.next.key, dummy);
        dummy.next = head.next;
    }
}

class Node{
     int key;
     int val;
     Node next;
    public Node(int key, int val){
        this.key = key;
        this.val = val;
        this.next = null;
    }
}

File: C#/146 LRU Cache.cs
public class LRUCache {
    int size;
    ListNode dummy;
    ListNode tail;
    IDictionary<int, ListNode> keyToPrev;

    public LRUCache(int capacity) {
        this.size = capacity;
        this.keyToPrev = new Dictionary<int, ListNode>();
        this.dummy = new ListNode(-1,-1);
        this.tail = this.dummy;
    }
    
    public int Get(int key) {
        if(!keyToPrev.ContainsKey(key)){
            return -1;
        }
        kick(key);
        return tail.val;
    }
    
    public void Put(int key, int value) {
        if(keyToPrev.ContainsKey(key)){
            kick(key);
            tail.val = value;
        }else{
            pushBack(new ListNode(key, value));
            if(keyToPrev.Count > size){
                popFront();
            }
        }
    }

    private void pushBack(ListNode node){
        keyToPrev.Add(node.key, tail);
        tail.next = node;
        tail = node;
    }

    private void kick(int key){
        ListNode prevNode = keyToPrev[key];
        ListNode keyNode = prevNode.next;
        if(keyNode == tail){
            return;
        }
        keyToPrev.Remove(keyNode.key);
        prevNode.next = keyNode.next;
        keyToPrev[prevNode.next.key] = prevNode;
        keyNode.next = null;
        pushBack(keyNode);
    }

    private void popFront(){
        ListNode head = dummy.next;
        keyToPrev.Remove(head.key);
        keyToPrev[head.next.key] = dummy;
        dummy.next = head.next;
    }
}

public class ListNode{
    public int key { get; set; }
    public int val { get; set; }
    public ListNode next { get; set; }
    public ListNode(int key, int val)
    {
        this.key = key;
        this.val = val;
    }
}